llamafile_sgemm API - INT8 implementation

amritahs-ibm opened the pull request:
This change upstreams llamafile's cpu matrix
multiplication kernels for ppc64le using MMA
builtins for quantised int8 datatype.

This change results in 10%-70% improvement
in total speed(ie all tokens/total time), across 
various batch sizes.

The patch is tested with Meta-Lllama-3-8B,
Mistral-7B, Llama-2-7B-chat-hf models on a
IBM POWER10 machine.

gcc-toolset-13 is minimum requirement to build
this patch. Run cmake config and build after path
has been updated with gcc-13:
export PATH=/opt/rh/gcc-toolset-13/root/usr/bin/:$PATH

amritahs-ibm committed:
llamafile_sgemm API - INT8 implementation

This change upstreams llamafile's cpu matrix
multiplication kernels for ppc64le using MMA
builtins for quantised int8 datatype.

This change results in 10% - 70% improvement
in total speed(ie all tokens/total time), across
various batch sizes.

The patch is tested with Meta-Lllama-3-8B,
Mistral-7B, Llama-2-7B-chat-hf models on a
IBM POWER10 machine.

Signed-off-by: Amrita H S <amritahs@linux.vnet.ibm.com>

amritahs-ibm commented:
Hi @ggerganov,
Can you please help reviewing this PR. Or suggest any missing actions required from me to get this patch reviewed.

slaren commented:
We will need to merge #10714 first, since there may be some conflicts.

amritahs-ibm commented:
Sure. Please let me know once that PR is merged. I will fix mine, if any conflicts and resubmit my PR.

Djip007 commented:
I'll try to submit it today. :crossed_fingers:

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1803,6 +2365,18 @@ bool llamafile_sgemm(int64_t m, int64_t n, int64_t k, const void *A, int64_t lda
+            k, (const block_q8_0 *)A, lda,
+            (const block_q8_0 *)B, ldb,
+            (float *)C, ldc,
+            ith, nth};
```
with https://github.com/ggerganov/llama.cpp/pull/10714

You will have to take ith/nth from params->ith/params->ith. It may be the only change you need to do.

amritahs-ibm replied:
I have made these changes and pushed. @slaren / @Djip007 / @ggerganov Please review the changes.

amritahs-ibm commented:
I have made the changes suggested by @Djip007 and pushed. @slaren / @Djip007 / @ggerganov Please review the changes.

Djip007 reviewed (commented):
These are quick personal comments wait for @slaren @ggerganov before change.
And I read it very quickly.

amritahs-ibm commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+        free(comparray);
+    }
+
+    void gemm_small(int64_t m0, int64_t m, int64_t n0, int64_t n, int RM, int RN) {
```
@Djip007 
I intend to use
std::array<int, 4> comparray; 
inside gemm_small as well. 

So is it necessary to make gemm_small template?. I am just thinking if making gemm_small a template function would have a minor performance impact. 

Please suggest.

Djip007 replied:
with RM/RN as template param the for is more simple for compiler to unroll or vectorise 

Sometime I use https://godbolt.org/ to see what compiler can do.

amritahs-ibm replied:
Done

amritahs-ibm commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+                    }
+                }
+                __builtin_mma_disassemble_acc(vec_C, &acc_0);
+                TA *aoffset = const_cast<TA*>(A+(ii*lda)+l);
```
@Djip007 
When I try to cast it normally like this:
 TA *aoffset = (TA*)(A+(ii*lda)+l);
 I get this warning:
 warning: cast from type ‘const block_q8_0*’ to type ‘block_q8_0*’ casts away qualifiers [-Wcast-qual].
 
 Is there any other way to avoid this warning?

Djip007 replied:
I was thinking more about
```c++
const TA* aoffset = A+(ii*lda)+l;
```
ie: no need to remove the const. (and not a good idea for compiler optimisation)

```c++
auto aoffset = A + ii*lda + l;
```

amritahs-ibm replied:
Done

amritahs-ibm commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+            mc = 8;
+            nc = 8;
+            gemm<8,8>(m0, m, n0, n);
+        } else if (m_rem >= 8 && n_rem >= 8) {
```
Yes, I had gemm<16,8> and gemm<8,16> earlier. But those had some performance problems and I switched to gemm<8,8>. 

I will remove that first 2 ifs now, and add later it later if I can solve the performance issues.

amritahs-ibm replied:
Done.

amritahs-ibm commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+        vec_t vec_A[8], vec_B[16] = {0};
+        vector signed int vec_C[4];
+        acc_t acc_0, acc_1;
+        int *comparray = (int *)malloc(4 * sizeof(int));
```
I have used std::array<int, 4> comparray; and remove free in all the kernels.

amritahs-ibm commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
       } \
    } \
 
+#define COMPUTE(ACC, c_idx, s_idx) \
+    __builtin_mma_disassemble_acc(vec_C, ACC); \
+    for (int i = 0; i < 4; i++) { \
+        CA[i] = vec_splats((float)(((double)comparray[c_idx+i]) * -128.0)); \
+        res[i] = vec_add(vec_ctf(vec_C[i], 0), CA[i]); \
+        fin_res[s_idx+i] = vec_madd(res[i], vs[s_idx+i], fin_res[s_idx+i]); \
+    } \
+
+#define SAVE_RES(ii, jj, RM, RN, idx) \
+    for (int I = 0; I < RM; I++) { \
+        for (int J = 0; J < RN; J++) { \
+            *((float*)(C+ii+((jj+J)*ldc)+I)) = *((float*)&fin_res[idx+I]+J); \
+        } \
+    } \
+
+#define PERMUTE_VEC(IR1, IR2, IR3, IR4, vecOffset) \
+    t1 = vec_perm(IR1, IR2, swiz1); \
+    t2 = vec_perm(IR1, IR2, swiz2); \
+    t3 = vec_perm(IR3, IR4, swiz1); \
+    t4 = vec_perm(IR3, IR4, swiz2); \
+    t5 = vec_perm(t1, t3, swiz3); \
+    t6 = vec_perm(t1, t3, swiz4); \
+    t7 = vec_perm(t2, t4, swiz3); \
+    t8 = vec_perm(t2, t4, swiz4); \
+    if (flip == true) { \
+        t5 = vec_xor(t5, xor_vector); \
+        t6 = vec_xor(t6, xor_vector); \
+        t7 = vec_xor(t7, xor_vector); \
+        t8 = vec_xor(t8, xor_vector); \
+    } \
+    vec_xst(t5, 0, vecOffset); \
+    vec_xst(t6, 0, vecOffset+16); \
+    vec_xst(t7, 0, vecOffset+32); \
+    vec_xst(t8, 0, vecOffset+48); \
```
@Djip007 
I used macros just to avoid some repitition. I can replace them by function or templates. 
From the performance viewpoint and considering the nature of the above macros, would function be better than templates?

Djip007 replied:
for example:
```c++
#define SAVE_RES(ii, jj, RM, RN, idx) \
    for (int I = 0; I < RM; I++) { \
        for (int J = 0; J < RN; J++) { \
            *((float*)(C+ii+((jj+J)*ldc)+I)) = *((float*)&fin_res[idx+I]+J); \
        } \
    } \
```
I will use template for RM/RN it make more likely let the compiler unrool the 2 loops for small values.
It may be sometime needed for various TYPE.
But yes, no need for template for most case.

amritahs-ibm replied:
Done

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+            int64_t ii = m0 + job / xtiles * RM;
+            int64_t jj = n0 + job % xtiles * RN;
+            //int *comparray = (int *)malloc(RM * sizeof(int), 4096);
+            int *comparray = (int *)malloc(RM * sizeof(int));
```
yes , if 
```c++
template<int RM, int RN>
void gemm_small(int64_t m0, int64_t m, int64_t n0, int64_t n)
```
then you can use
```c++
std::array<int, RM> comparray;
```

amritahs-ibm replied:
Done

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+        int64_t duty = (tiles + nth - 1) / nth;
+        int64_t start = duty * ith;
+        int64_t end = start + duty;
+        if (RM == 4 && RN == 8) {
+            kernel = &tinyBLAS_Q0_PPC::KERNEL_4x8;
+        } else if (RM == 8 && RN == 4) {
+            kernel = &tinyBLAS_Q0_PPC::KERNEL_8x4;
+        } else if (RM == 8 && RN == 8) {
+            kernel = &tinyBLAS_Q0_PPC::KERNEL_8x8;
+        }
```
In this case, I would have tried to make KERNEL_N_M a template, if not possible/simple I would have done
something like that now we can use c++17:
```c++
tempate<int RN, int RM>
inline void kernel(int64_t ii, int64_t jj) {
    if constexpr(RM == 4 && RN == 8) {
        KERNEL_4x8(ii,jj);
    } else if constexpr(RM == 8 && RN == 4) {
        KERNEL_4x8(ii,jj);
    } else if constexpr(RM == 8 && RN == 8) {
        KERNEL_8x8(ii,jj);
    } else {
        static_assert(false, "RN/RM values not supported");
    }
}
```
and use
```c++
    kernel<RM,RN>(ii, jj);
```
This remove the need of "methode-pointer", and do not compile if you use it with not supported RM/RN.

amritahs-ibm replied:
Done

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1049,577 @@ class tinyBLAS_Q0_AVX {
+    }
+
+  private:
+    void (tinyBLAS_Q0_PPC::*kernel)(int64_t, int64_t);
```
Did not see it last time, dangerous, it is not thread safe.

amritahs-ibm replied:
Removed.

amritahs-ibm commented:
All comments are addressed expect for the last MMA one. The updated patch has been committed.
I will look into MMA comment and get back to you.

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1050,704 @@ class tinyBLAS_Q0_AVX {
+  private:
+
+    template<int RM, int RN>
+    void save_res(int ii, int jj, int idx, vector float* fin_res) {
```
You can add inline .  It may (or not) give some optimised code.

amritahs-ibm replied:
Done

Djip007 commented on ggml/src/ggml-cpu/llamafile/sgemm.cpp:
```diff
@@ -1049,6 +1050,704 @@ class tinyBLAS_Q0_AVX {
+    }
+
+    template<int size>
+    void compute(acc_t* ACC, int c_idx, int s_idx, std::array<int, size>& comparray, vector float* vs, vector float* fin_res) {
```
You can add inline to.

amritahs-ibm replied:
Done

amritahs-ibm commented:
> @amritahs-ibm I didn't realize that MMA was "Matrix Multiply Accelerate". Have you see what was done with amx or "aarch64" kernel? There is now "simple" arch to create kernel that can "repack" the weight, so it fit to the structure needed for such Matrix OP? That way A is repack at load time once, and only B need to be repack at runtime.

Are you referring to gemm4xN and gemmMx4 functions in tinyBLAS_Q0_AVX?

Also in case of PowerPC's MMA for int8 data type, MMA engine requires the data to be packed in a different way. So I came up with a specific function for int8(ie packNormal) to do the packing.
![image](https://github.com/user-attachments/assets/981f90e1-5e81-46c7-953b-f58bd1a584fe)

Please find below the MMA guide:
https://www.redbooks.ibm.com/redpapers/pdfs/redp5612.pdf

amritahs-ibm commented:
@Djip007 Could you please help me on my latest comment?

amritahs-ibm commented:
I have addressed all the comments except for the MMA repacking one. As it is an optimization over the current code, I can take it up in the follow on patch. 
@slaren @Djip007 @ggerganov 
Can anyone of you approve the patch and get it merged?
